components/blitz/src/omero/gateway/facility/ROIFacility.java

```diff
@@ -834,71 +846,190 @@ else if (shape.isDirty()) {
+                                .id(imageId).childType("Roi")
+                                .stopBefore("Shape").build();
+
+                        cb = gateway.submit(ctx, finder);
+                        cb.block(10000);
+                        found = (FoundChildren) cb.getResponse();
+                        List<Long> imageRoiIDs = found.children
+                                .remove(ome.model.roi.Roi.class.getName());
+
+                        Iterator<Long> it = roiIDs.iterator();
+                        while (it.hasNext()) {
+                            if (!imageRoiIDs.contains(it.next()))
+                                it.remove();
+                        }
+                    } else {
+                        // only take ROIs into account which are not attached to
+                        // an image
+                        IQueryPrx service = gateway.getQueryService(ctx);
+                        ParametersI p = new ParametersI();
+                        p.addIds(roiIDs);
+
+                        String query = "select distinct roi from Roi roi "
+                                + "where roi.image is null "
+                                + "and roi.id in (:ids) ";
+
+                        List<IObject> objs = service.findAllByQuery(query, p);
+
+                        roiIDs.clear();
+                        for (IObject obj : objs) {
+                            Roi roi = (Roi) obj;
+                            roiIDs.add(roi.getId().getValue());
+                        }
+                    }
+                }
             }
 
-            // filter the ROIResults
-            Iterator<ROIResult> it = roiresults.iterator();
-            while (it.hasNext()) {
-                ROIResult r = it.next();
-                Iterator<ROIData> it2 = r.getROIs().iterator();
-                while (it2.hasNext()) {
-                    ROIData roi = it2.next();
-                    if (!roiIds.contains(roi.getId()))
-                        it2.remove();
-                }
+            return loadROIs(ctx, roiIDs);
+        } catch (Throwable e) {
+            handleException(this, e, "Cannot load ROIs for folders ");
+        }
 
-                if (r.getROIs().isEmpty())
-                    it.remove();
-            }
+        return Collections.EMPTY_LIST;
+    }
 
-            return roiresults;
+    /**
+     * Load all ROIs with the given ids
+     * 
+     * @param ctx
+     *            The {@link SecurityContext}
+     * @param ids
+     *            The ROI ids
+     * @return See above.
+     * @throws DSOutOfServiceException
+     * @throws DSAccessException
+     */
+    public Collection<ROIData> loadROIs(SecurityContext ctx,
+            Collection<Long> ids) throws DSOutOfServiceException,
+            DSAccessException {
+        try {
+            if (ids == null || ids.isEmpty())
+                return Collections.EMPTY_LIST;
```

Comment:
if you're getting a yellow Eclipse warning on this you should find that you can fix it with `Collections.emptyList()`